insecure/corruptible/factory.go

```diff
@@ -142,13 +209,14 @@ func (c *ConduitFactory) registerAttacker(ctx context.Context, address string) e
 		return fmt.Errorf("illegal state: trying to register an attacker (%s) while one already exists", address)
 	}
 
-	clientConn, err := grpc.Dial(address)
+	clientConn, err := grpc.Dial(address,
+		grpc.WithTransportCredentials(grpcinsecure.NewCredentials()))
 	if err != nil {
 		return fmt.Errorf("could not establish a client connection to attacker: %w", err)
 	}
 
 	attackerClient := insecure.NewAttackerClient(clientConn)
-	c.attackerObserveClient, err = attackerClient.Observe(ctx)
+	c.attackerObserveClient, err = attackerClient.Observe(c.ctx)
```

Comment:
why use the factory's context instead of the one passed in? generally, it's best to [avoid storing contexts](https://go.dev/blog/context-and-structs)

Reply:
The short answer is: to guarantee a graceful shutdown and avoid message loss. 


The long answer is: this is a `gRPC` method invoked by a remote attacker. As the result of invocation, it tries bouncing back a `gRPC` connection to the attacker (i.e., `attackerClient`). The input context is coming from the _remote_ attacker's side and is influenced by the attacker's lifecycle, for the rest of this comment, let's call it **the attacker's context**. 

Now if we use the incoming attacker's context to create a `gRPC` connection back to the attacker, we are making the lifetime of this connection totally coupled with the attacker's life cycle. This poses an issue: when the attacker shutdown _starts_ by its component manager, the contexts generated at the attacker's side are getting canceled right away. The cancellation of context should gracefully stop the attack network by first letting messages en route on connections to be delivered, and then tearing down the connections. 

However, when we use attacker's context to create a connection back to it, the resulted connection will be canceled the moment that the parent context at the attacker side is canceled and _before_ the `gRPC` server at the attacker's side shuts down gracefully. So, there is a chance of message loss. 

This message loss is verifiable by switching the code above to use the passed `ctx` and then the test gets flakey since the moment attacker starts shutting down, all connections are dropped immediately and messages en route will be dropped. 

Accordingly, the context to create connections back to attacker should be local to the `factory` and managed by it. The simplest solution is to create a `context.Background()` every time  `factory` is creating a connection to the attacker, i.e., 

```
c.attackerObserveClient, err = attackerClient.Observe(context.Background())
```
However, the problem is those connections may outlive the `factory` itself, since they may reside even when the factory itself is shut down. That is the reason behind having a local context as a member field to the `factory` and using it to create connections back to the attacker, hence making sure that: 
1. Those connections are not controlled by the attacker's lifecycle. 
2. Those connections are torn down as the procedure of `factory`'s shutdown. 

P.S: when the attacker is shut down, the `gRPC` connection is closed at the attacker side, and when all messages are delivered, an `EOF` is delivered on the connection telling the `factory` that the connection is over.